### PR TITLE
[WIP] sql: define a first version of plan rewinding

### PR DESCRIPTION
**Pull request by knz**
Note: all commits but the last belong to #21305. Will rebase when that merges.

tl;dr: this patch is a 70% solution to enable reuse of logical plans
(planNode hierarchies) for multiple executions.

Prior to this patch, the planNode interface had been designed with the
assumption that a planNode tree would only be executed once, and
subsequently close()d. This assumption expresses itself in multiple
code paths that modify planNodes in-place and "consume" them (destroy
information), effectively preventing multiple reuse.

This is an obstacle to the caching of logical plans.

This patch advances the roadmap toward plan caching by changing the
interface and adding infrastructure so that *some* plans can be
"rewound" and brought back to a point where they can be executed
again.

This is achieved as follows:

- a predicate `isRewindable()` is defined, aiming to return true iff a
  plan can be reused multiple times.

  Its existence is motivated by the assumption that not every plan can
  actually be reused. In fact, during the implementation of this
  patch I encountered significant hurdles trying to make DML
  statements and the windowNode rewindable, and this interface allows
  me to side-step these hurdles and release this patch with support
  for the most common read-only queries.

- nodes that qualify as "rewindable" by `isRewindable()` can be either
  implicitly rewound (i.e. there is nothing particular to do other
  than call `startExec` again before restarting execution), or
  *explicitly* rewound where a new method `rewindExec` must be called.

  A top-level function `(*planTop).restart()` and accompanying
  `restartPlan()` is introduced that automate the call to `rewindExec`
  across a planNode tree.

This patch, once complete, will be merely a "70% solution" because:

- not every planNode can be rewound.
- this approach repeats initialization of some fields between the
  planNode constructor and the `rewindExec` method. A more future
  proof (but more costly in terms of engineering time) would be to
  factor the initialization of data structures into a single place.
- any schema descriptors captured in a planNode are left captured and
  not re-initialized/re-read upon rewind, thereby kicking the can of
  plan invalidation upon schema updates down the road.

Nevertheless, a "70% solution" should be considered acceptable at this
point because it unlocks further progress on the implementation of
plan caching.

----

Work in progress--the following concerns must be addressed before this
patch is complete:

- the rewinding must be extended to support common node types
  which we'd really want to support even in the first version
  (e.g. scans!)
- the rewinding must be extended to support subqueries.
- the distsql physical planning must be audited to ensure that it does
  not modify planNodes in-place and can be run multiple times.

### CODE REVIEW

**Comment by knz** [comment]
@jordanlewis here we are, hopefully you can start building from this.

I'd like us to spend a little time together to look at `scanNode` at least, and possibly the other few that still need some attention. Since you're paged in on the row fetcher, I'd like to pick your brain as to how to best re-initialize that during rewinds.

**Comment by knz** [comment]
Also we need to make progress on #21305 before I can complete this patch. Unfortunately some commit last week broke that PR and I am unable to resolve the conflict. I asked Radu to have a look for the last conflict but really this is bound to conflict again every week until it merges.

**Comment by jordanlewis** [comment]
:lgtm:

I wish it were easier! It would be nice if this rewind method could just basically say plan.Run = myPlanRunStruct{}, zeroing the fields. It seems like the memory monitors are the problem with that strategy, right? Perhaps they deserve to be put into a different struct, or something like that.

Maybe the memory monitors should be closed in a separate method that gets called at the end of plan execution (and in Close()) instead of during reset.

Also, how do we test this? Can we hook into logic tests somehow (for r-o statements), rewind the plan, and assert that the result is the same?

---

Reviewed 2 of 6 files at r5, 1 of 2 files at r6, 1 of 2 files at r7, 5 of 6 files at r9, 20 of 20 files at r10, 21 of 21 files at r11.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/plan.go, line 429 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/21446#-L2v-U8fwAi5GOMQY_Fn:-L2v-U8fwAi5GOMQY_Fo:b56aonr) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b9e88ab6d8c269bfce801f7394e571085802252/pkg/sql/plan.go#L429)):*
> ```Go
> 		},
> 		leaveNode: func(_ string, n planNode) error {
> 			if rewind {
> ```

Can't help but notice that we have the opportunity to substitute an entirely new function here, rather than putting the conditional inside the function. But it likely doesn't matter here because this function won't get called often or in the hot path.

---

*[pkg/sql/rewind.go, line 133 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/21446#-L2v-MUmJLsTyS8XipgR:-L2v-MUmJLsTyS8XipgS:b-g478cy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b9e88ab6d8c269bfce801f7394e571085802252/pkg/sql/rewind.go#L133)):*
> ```Go
> // splitNode is rewindable.
> 
> // FIXME(knz/joran) make unionNode rewindable. This may be complicated
> ```

s/jordan/jordan/ :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21446#-:-L2v-tckLR8sogQDWinv:b9e1wbx)*

**Comment by knz** [comment]
> Also, how do we test this? Can we hook into logic tests somehow (for r-o statements), rewind the plan, and assert that the result is the same?

I wanted to append a commit that adds a test knob in the executor and makes it run the plan once, discard the result, rewind the plan and execute again, the 2nd time delivering the results to the client. 

Let that commit validate CI

Then append another commit that removes the test knob.

What do you think?
